Connect to services over dmsghttp

ersonp opened the pull request:
Did you run `make format && make check`?
yes

Fixes #985 

 Changes:	
- Updated `disc.NewHTTP`
- Added `http.Client` param to `dmsgc.New()`
- Update `initDmsg()` to connect to the `dmsg-discovery` via HTTP over dmsg if the URL of `dmsg-discovery` in the config is `dmsg://<pk-of-dmsg-discovery>:80`
- Added `dmshhttp` module
- Renamed `localServers.json` to `dmsghttp-config.json`
- Changed datatype of `DMSGServers` from `[]string` to `[]*disc.Entry`

How to test this PR:

On test deployment:
1. Run
    ```
    ./skywire-cli config gen -dirt
    ```
2. Run 
    ```
    ./skywire-visor skywire-config.json
    ```
3. Open hypervisor UI and check functionality. 

On integration env:
Check [#skywire-services-486](https://github.com/SkycoinPro/skywire-services/pull/486)

Needs:
[#dmsg-130](https://github.com/skycoin/dmsg/pull/130)

ersonp committed:
Add dmsg direct client

The commit adds dmsg direct client for communication with services via http over dmsg.

ersonp committed:
Fix StartDmsg

ersonp committed:
Add dmsgHTTPC to router

The commit adds dmsgHTTPC to the router that enables the visor to connect to the route-finder via http over dmsg.

ersonp committed:
Add dmsgHTTPC to router

The commit adds dmsgHTTPC to the router that enables the visor to connect to the route-finder via http over dmsg.

ersonp committed:
Fix datarace on shutdown

ersonp committed:
Add dmsghttp to transport-discovery

The commit adds the functionality for the tpdclient to connect to the transport-discovery via http over dmsg. It also contains changes to httpauth to accommodate this functionality.

ersonp committed:
Rename dmsgHTTPC to httpC

The commit renames dmsgHTTPC to httpC as it better reflects it's use as it can either be a dmsghttp client or just a http client.

ersonp committed:
Add dmsghttp to uptime-tracker

The commit adds the functionality for the utclient to connect to the uptime-tracker via http over dmsg.

ersonp committed:
Add dmsghttp to uptime-tracker

The commit adds the functionality for the utclient to connect to the uptime-tracker via http over dmsg.

ersonp committed:
Add dmsghttp to service-discovery

The commit adds the functionality for the appdisc and servicedisc to connect to the service-discovery via http over dmsg.

ersonp committed:
Add dmsghttp to public autoconnect

The commit adds the functionality for the public autoconnect to connect to the service-discovery via http over dmsg.

ersonp committed:
Add dmsghttp to address-resolver

The commit adds the functionality for the address-resolver client to connect to the address-resolver via http over dmsg.

ersonp committed:
Minor changes

The commit changes the var disc to serviceURL and also changes the src string of pushCloseStack in initUptimeTracker to uptime_tracker.

ersonp committed:
Fix service-disc url

ersonp committed:
Merge branch 'develop' into feature/dmsghttp

ersonp committed:
Update locakServers.json

ersonp committed:
Change Dmsg Servers datatype

This commit changes the datatype of dmsg servers from []string to []*disc.Entry to better suit it's usage.

ersonp committed:
Add dmsghttp module

This commit adds dmsghttp module moving the older repetative code to one func. It also incudes the func getHTTPClient which further streamlines the code.

The service modules address-resolver, discovery, transport, dmsg, router and uptime_tracker are are reliant on the new module.

The dClient (dmsg direct client) and dmsgHTTP (dmsghttp client) are added to the Visor struct.

mrpalide reviewed (approved):
Code is clean and actually work both test, in integration env and locally independent.

jdknives commented on pkg/dmsgc/dmsgc.go:
```diff
@@ -13,13 +14,13 @@ import (
 
 // New makes new dmsg client from configuration
-func New(pk cipher.PubKey, sk cipher.SecKey, eb *appevent.Broadcaster, conf *DmsgConfig) *dmsg.Client {
+func New(pk cipher.PubKey, sk cipher.SecKey, eb *appevent.Broadcaster, conf *DmsgConfig, httpC http.Client) (dmsgC *dmsg.Client) {
```
Why the named return here?

ersonp replied:
Changed it.

ersonp committed:
Requested changes

ersonp committed:
Update dmsg

ersonp committed:
Rename localServers.json

This commit renames the config file localServers.json to dmsghttp-config.json along with it's reference in config gen.

ersonp committed:
Merge branch 'develop' into feature/dmsghttp

ersonp committed:
Change data type of DMSGServers

This commit changes the data type of DMSGServers from []string to []*disc.Entry.

ersonp committed:
Update changelog

jdknives commented on cmd/skywire-cli/commands/config/gen.go:
```diff
@@ -107,7 +107,7 @@ var genConfigCmd = &cobra.Command{
-			serversListJSON, err := ioutil.ReadFile("localServers.json")
+			serversListJSON, err := ioutil.ReadFile("dmsghttp-config.json")
 			if err != nil {
 				logger.WithError(err).Fatal("Failed to read servers.json file.")
```
Comment needs to be updated.

jdknives reviewed (approved):
Looks good. Just one small point.

ersonp committed:
Fix comment
